app/src/main/kotlin/compiler/intermediate_form/ControlFlowGraphBuilder.kt

```diff
@@ -0,0 +1,70 @@
+package compiler.intermediate_form
+
+import compiler.common.reference_collections.ReferenceHashMap
+import compiler.common.reference_collections.referenceHashMapOf
+
+class ControlFlowGraphBuilder(var entryTreeRoot: IFTNode? = null) {
+    private var unconditionalLinks = ReferenceHashMap<IFTNode, IFTNode>()
+    private var conditionalTrueLinks = ReferenceHashMap<IFTNode, IFTNode>()
+    private var conditionalFalseLinks = ReferenceHashMap<IFTNode, IFTNode>()
+    private var treeRoots = ArrayList<IFTNode>()
+
+    init {
+        if (entryTreeRoot != null)
+            treeRoots.add(entryTreeRoot!!)
+    }
+
+    fun addLink(from: Pair<IFTNode, CFGLinkType>?, to: IFTNode) {
+        if (from != null) {
+            for (node in listOf(from.first, to))
+                if (!treeRoots.contains(node))
+                    treeRoots.add(node)
+
+            val links = when (from.second) {
+                CFGLinkType.UNCONDITIONAL -> unconditionalLinks
+                CFGLinkType.CONDITIONAL_TRUE -> conditionalTrueLinks
+                CFGLinkType.CONDITIONAL_FALSE -> conditionalFalseLinks
```

Comment:
Shouldn't we verify somewhere that these links have corresponding pairs?

Reply:
Added verification when calling `build` and it turns out that 2 unit tests (namely `unreachable statement because of continuation` and `function return`) in `FunctionControlFlowTest` are failing - there are true links without matching false links and vice versa. Temporarily I disabled these tests, but we should take a look at them with high priority.